allen-cell-animated/nucmorph-colorizer
# fix: Fixed bug where dropdown would render over modals

Problem
=======
*Estimated review size: small, 10 minutes*

Closes #276, a bug where the color map dropdown would render over the modals.

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/97925f11-9e2f-4cd1-926a-130e76d0e11f)
*Horrifying!*

Solution
========
The issue was with the modals themselves. They were being contained within a div along with the buttons, which meant that they were in a [**stacking context**](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context) with the buttons that prevented them from overlaying the rest of the UI. (Elements in a stacking context can apply z-indexing to change their ordering ONLY within the context.)

Removing the container also didn't work because it causes them to live outside of the styles and CSS variables given by `AppStyle`.

The solution was to add a container reference in `AppStyle` that can be  accessed through React's `Context` API. This means that modals can still be styled using rules defined in `AppStyle`, but aren't confined to the z-sorting of their root triggering buttons.

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/bfde2cbc-0964-410a-8540-6e528eff6659) 

### Changes:

- Add a `DocumentContext` containing an `HTMLDivElement` ref to a modal container.
- Allow `AppStyle` to provide a value for `DocumentContext`'s modal container ref.
- Update the existing modals to use the new context and ref.

## Type of change
* Bug fix (non-breaking change which fixes an issue)


Steps to Verify:
----------------
1. Open the preview link: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-280/
2. Click on the color ramp dropdown button to pin it open.
3. Quickly click the Load or Export buttons. The color ramp dropdown should appear behind the modal.

Screenshots (optional):
-----------------------
| Before | After |
| --- | --- |
| ![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/97925f11-9e2f-4cd1-926a-130e76d0e11f) | ![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/b3f6d58f-7b86-414a-9957-1113d31d4151) |
